odoo/orm/domains.py

```diff
@@ -891,6 +891,8 @@ def _optimize_field_search_method(self, model: BaseModel) -> Domain:
         ))
 
     def _to_sql(self, model: BaseModel, alias: str, query: Query) -> SQL:
+        field = self._field(model)
+        model._check_field_access(field, 'read')
```

Comment:
I am not sure where we should check the field access. `BaseModel._field_to_sql` still checks field access, but condition_to_sql does not do it anymore - I suppose the reason is that it is just a hook (called mainly here).

We could even check this in the `_optimize` method. It would check permission even on non-sqlable fields.

Reply:
> We could even check this in the _optimize method. It would check permission even on non-sqlable fields.

That would indeed make sense :thinking:

Reply:
By the way, I believe that, as usual in Odoo, security checks are at the wrong places.  By clearly separating the security responsibility from the functionality, it makes it easier to implement something like "bypass security up to a certain point, but not beyond", which is what we need with `related_sudo` fields.